### PR TITLE
PWN-3806 - Add feature toggles in app

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-3806

## Description of Work

- add feature toggle entity with subclasses that declare concrete types
- edit feature toggles settings screen
- add RemoteConfig fetch from Firebase

## Screenshots (Optional)
Screen Shot 2022-06-22 at 15 24 54

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/debug/feature_toggles/FeatureTogglesAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            return oltItem.toggleName == newItem.toggleName && oltItem.toggleValue == newItem.toggleValue
+        }
+
+        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+            return oldList[oldItemPosition] == newList[newItemPosition]
+        }
+    }
+
+    // not so much data for using DiffUtil
+    @SuppressLint("NotifyDataSetChanged")
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/debug/feature_toggles/FeatureTogglesAdapter.kt:43]: do we need this supress?

File: app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt [hunk 2]
```diff
@@ -7,7 +7,7 @@ import org.koin.android.ext.android.inject
 import org.p2p.wallet.R
 import org.p2p.wallet.common.mvp.BaseMvpFragment
 import org.p2p.wallet.databinding.FragmentDebugSettingsBinding
-import org.p2p.wallet.debug.featuretoggles.FeatureTogglesFragment
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt:10]: do not like _ in packages

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt:10]: hm, i thought we have set naming in our code style, but i was wrong.
should we use Kotlin `camelCase.packageName` style then?

Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt:10]: @1sakov ^^ ?
i prefer camelCase if this won't harm our CI for example

Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt:10]: I'd leave as it is. 
We have several similar folders already. like pushnotifications

So, let's leave as `featuretoggles`

**Review by 1sakov — CHANGES_REQUESTED** [review]
Added comments. 
In general LGTM, it's extendable when new feature is added and I like it. 

But I'm confused what class should I use when I want to hide it behind feature toggle? 
There are different classes can be injected and imho this is bad

File: app/src/main/java/org/p2p/wallet/common/feature_toggles/FeatureToggle.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+abstract class FloatFeatureToggle(
+    private val valuesProvider: RemoteConfigValuesSource
+) : FeatureToggle<Float>() {
+    override val value: Float get() = valuesProvider.getFloat(toggleKey) ?: defaultValue
+}
+
+/**
+ * Use subclasses to override how should be JSON parsed
+ */
+abstract class JsonFeatureToggle<GSON_CLASS>(
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/common/feature_toggles/FeatureToggle.kt:47]: may name it's as GsonInstance or GsonClass

**Review by davran1001 — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/debug/settings/DebugSettingsFragment.kt:10]: done

File: app/src/main/java/org/p2p/wallet/debug/feature_toggles/FeatureTogglesAdapter.kt (hunk 1 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/debug/feature_toggles/FeatureTogglesAdapter.kt:43]: removed

File: app/src/main/java/org/p2p/wallet/debug/feature_toggles/FeatureTogglesAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+        lateinit var currentRow: FeatureToggleRow
+
+        private val toggleValueInputWatcher = object : TextWatcher {
+            override fun onTextChanged(text: CharSequence?, start: Int, before: Int, count: Int) {
+                text?.let { onToggleChanged.invoke(currentRow, it.toString()) }
+            }
+
+            override fun afterTextChanged(s: Editable?) = NoOp
+            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = NoOp
```
[the remark this answers is not in the document]
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/debug/feature_toggles/FeatureTogglesAdapter.kt:72]: Let's discuss this on the sync

**Review by 1sakov — APPROVED** [review]

**Comment by gslevinkov** [comment]
check hook

**Review by kuznetsov — APPROVED** [review]

**Pull request merged**